### PR TITLE
Upsert resource improvements

### PR DESCRIPTION
**Pull request by phanimarupaka**
This PR addresses issue https://github.com/GoogleContainerTools/kpt/issues/2570

We are enabling 2 new functionalities with this PR:

1. upsert-resource function will be accepting new Custom Resource `UpsertResourceList`. Users can wrap multiple resources to be upserted and the function will unwrap them and upsert them all. With this, the users need not invoke the function once per each upsert operation which cuts down latency and network calls.
2. We introduce a new annotation `config.kubernetes.io/upsert-path` which when specified in the input resource to be upserted, this function makes sure that the resource is upserted in that specified path instead of the top level directory.

### CODE REVIEW

**Comment by phanimarupaka** [comment]
@mengqiy Can you please check as to why is checkSite failing?

File: examples/upsert-resource-multiple/.fn-config/fn-config.yaml
Comment (droot) [inline examples/upsert-resource-multiple/.fn-config/fn-config.yaml:1]: I think we have apiVersion for all functions in catalog ..something starting with `fns.kpt.dev` or something.

pl. ensure it is consistent with other functionConfigs.

/cc @mengqiy

File: functions/go/upsert-resource/upsertresource/upsert_resource.go
Comment (droot) [inline functions/go/upsert-resource/upsertresource/upsert_resource.go:19]: hmm...I think it will be good to narrow the scope for annotations to this function. `config.kubernetes.io` is generic prefix being used in k8s ecosystem. @mengqiy any recommendation ?

File: examples/upsert-resource-at-path/.expected/config.yaml
Comment (droot) [inline examples/upsert-resource-at-path/.expected/config.yaml:1]: Do we have to specify this ? it should be idempotent by default.. right.

**Review by droot — COMMENTED** [review]
Looks good, I have minor comments.

File: examples/upsert-resource-at-path/.fn-config/fn-config.yaml
Comment (morgante) [inline examples/upsert-resource-at-path/.fn-config/fn-config.yaml:7]: I don't think this annotation should be named specifically for `upsert`, since if/when we add other functions for modifying resources directly (ex. `patch`) I'd expect them to use the same annotation.

Maybe `target-path`?

File: examples/upsert-resource-multiple/.fn-config/fn-config.yaml
Comment (morgante) [inline examples/upsert-resource-multiple/.fn-config/fn-config.yaml:2]: Why `UpsertResourceList` vs the existing `ResourceList`? Aren't the schemas identical?

File: functions/go/upsert-resource/upsertresource/upsert_resource.go
Comment (morgante) [reply functions/go/upsert-resource/upsertresource/upsert_resource.go:19]: I don't think this annotation should be named specifically for `upsert`, since if/when we add other functions for modifying resources directly (ex. `patch`) I'd expect them to use the same annotation.

Maybe `target-path`?

File: examples/upsert-resource-at-path/README.md
Comment (phanimarupaka) [reply examples/upsert-resource-at-path/README.md:13]: for this function it function config must come from outside of the package directory. It doesn't make sense to have the resource within the package and have it upserted by the function into package.

File: examples/upsert-resource-at-path/.expected/config.yaml
Comment (mengqiy) [reply examples/upsert-resource-at-path/.expected/config.yaml:1]: I don't think we need this file.

File: examples/upsert-resource-at-path/.expected/exec.sh
Comment (mengqiy) [inline examples/upsert-resource-at-path/.expected/exec.sh:4]: Our test harness supports eval and fn-config, why not using that?
Ref: https://github.com/GoogleContainerTools/kpt/tree/main/pkg/test/runner
```yaml
testType: eval
image: ...
fnConfig: ...
```

File: examples/upsert-resource-multiple/.expected/config.yaml
Comment (mengqiy) [inline examples/upsert-resource-multiple/.expected/config.yaml:1]: Same comment here.

File: examples/upsert-resource-multiple/.expected/exec.sh
Comment (mengqiy) [inline examples/upsert-resource-multiple/.expected/exec.sh:4]: Same comment here.

File: examples/upsert-resource-multiple/.fn-config/fn-config.yaml
Comment (mengqiy) [reply examples/upsert-resource-multiple/.fn-config/fn-config.yaml:1]: We are using `fn.kpt.dev/v1alpha1` for functions authored by the kpt team.

File: examples/upsert-resource-multiple/.fn-config/fn-config.yaml
Comment (mengqiy) [reply examples/upsert-resource-multiple/.fn-config/fn-config.yaml:2]: `UpsertResourceList` is identical to `v1.List` schema-wise. `ResourceList` has more fields than `v1.List`. e.g. `functionConfig` and `results`.
Another reason is that we want the function to work with both kpt and kustomize. To work with kustomize, a function must support a unique GVK as its functionConfig, since kustomize uses this GVK to identify functions.

When we build a new function we want to make it work with both kpt and kustomize. And it won't hurt the interoperability between kpt and kustomize.

I guess we can simply use `kpt fn source | sed` to get an `UpsertResourceList`.

File: functions/go/upsert-resource/upsertresource/upsert_resource.go
Comment (mengqiy) [reply functions/go/upsert-resource/upsertresource/upsert_resource.go:19]: in the current design, annotation `DestPathAnnotation` has the same meaning as `internal.config.kubernetes.io/path`. IMO I'd suggest using an annotation with kpt prefix, so users a less likely to confuse these 2 annotations.
Maybe we can consider something like `config.kpt-dev/target-path` or `fn.kpt-dev/target-path`.

File: examples/upsert-resource-multiple/.fn-config/fn-config.yaml
Comment (morgante) [reply examples/upsert-resource-multiple/.fn-config/fn-config.yaml:2]: > UpsertResourceList is identical to v1.List schema-wise. ResourceList has more fields than v1.List. e.g. functionConfig and results.

If that's the case, why don't we just use `v1.List`?

We should not proliferate separate Kinds for each function. If they have the same semantic meaning and schema, they should be the same Kind.

> Another reason is that we want the function to work with both kpt and kustomize. To work with kustomize, a function must support a unique GVK as its functionConfig, since kustomize uses this GVK to identify functions.

Other functions accept ConfigMaps generically, that seems strange. In general, I think this is a poor design choice we should reconsider.

File: examples/upsert-resource-multiple/.fn-config/fn-config.yaml
Comment (phanimarupaka) [reply examples/upsert-resource-multiple/.fn-config/fn-config.yaml]: We will be using List type, so it would be v1

File: functions/go/upsert-resource/upsertresource/upsert_resource.go
Comment (phanimarupaka) [reply functions/go/upsert-resource/upsertresource/upsert_resource.go]: we are going with config.kubernetes.io/target-path

File: examples/upsert-resource-multiple/.fn-config/fn-config.yaml
Comment (phanimarupaka) [reply examples/upsert-resource-multiple/.fn-config/fn-config.yaml]: We will be using v1.List.

File: examples/upsert-resource-at-path/.expected/exec.sh
Comment (phanimarupaka) [reply examples/upsert-resource-at-path/.expected/exec.sh]: Updated.

File: examples/upsert-resource-multiple/.expected/config.yaml
Comment (phanimarupaka) [reply examples/upsert-resource-multiple/.expected/config.yaml]: Removed

File: examples/upsert-resource-multiple/.expected/exec.sh
Comment (phanimarupaka) [reply examples/upsert-resource-multiple/.expected/exec.sh]: Updated.

**Comment by mengqiy** [comment]
You are working with path in the function. So can you please upgrade kyaml to 0.13.0 for this function?

File: examples/upsert-resource-multiple/README.md
Comment (mengqiy) [inline examples/upsert-resource-multiple/README.md:16]: > ofqq

?

File: examples/upsert-resource-multiple/README.md
Comment (mengqiy) [inline examples/upsert-resource-multiple/README.md:23]: > .expected/fn-config.yaml

`.expected` is used to host test harness related stuff while `fn-config.yaml` is user-facing content. It should live in another directory.

File: examples/upsert-resource-at-path/README.md
Comment (mengqiy) [inline examples/upsert-resource-at-path/README.md:22]: Same comment here

File: functions/go/upsert-resource/metadata.yaml
Comment (mengqiy) [inline functions/go/upsert-resource/metadata.yaml:7]: I'd suggest to rename it to `upsert-resource-single`

File: examples/upsert-resource-at-path/.expected/config.yaml
Comment (phanimarupaka) [reply examples/upsert-resource-at-path/.expected/config.yaml]: Removed.

File: examples/upsert-resource-multiple/README.md
Comment (phanimarupaka) [reply examples/upsert-resource-multiple/README.md]: Typo. Fixed it.

File: functions/go/upsert-resource/metadata.yaml
Comment (phanimarupaka) [reply functions/go/upsert-resource/metadata.yaml:7]: It is already out there. So better not rename it.

File: examples/upsert-resource-multiple/README.md
Comment (phanimarupaka) [reply examples/upsert-resource-multiple/README.md:23]: I feel fn-config is input to function and it should live next to `config.yaml` or `exec.sh`.

**Comment by phanimarupaka** [comment]
> You are working with path in the function. So can you please upgrade kyaml to v0.12.1-0.20211012224254-55ac9ca88db9 for this function?
> 
> EDIT: Please use kyaml v0.12.1-0.20211012224254-55ac9ca88db9

@mengqiy I tried that but it seems like we need a major refactor for this function to use that version of kyaml. And technically since the change is backwards compatible, beta.8+ of kpt should work with this function fine. And I verified that it is working fine. Keeping the timeline in view, lets push this change and then upgrade the kyaml version for this function.

File: examples/upsert-resource-multiple/.expected/config.yaml
Comment (mengqiy) [inline examples/upsert-resource-multiple/.expected/config.yaml:4]: This test can be unskiped as soon as we merge https://github.com/GoogleContainerTools/kpt-functions-catalog/pull/644/commits.

**Comment by phanimarupaka** [comment]
@mengqiy I will have a separate PR for that. Lets not block this PR for the other one.

**Review by mengqiy — APPROVED** [review]

**Comment by williamjiuchengcai** [comment]
I've verified these changes locally. They seem to be doing what I expect them to (I upserted 2 different setters.yaml files using a fn-config containing a resource list):

test function config:
```
apiVersion: v1
kind: List
metadata:
  name: upsert-multiple-resources-config
items:
- apiVersion: v1
  kind: ConfigMap
  metadata:
    name: setters
    annotations:
      config.kubernetes.io/target-path: 'setters.yaml'
  data:
    namespace: config-control-1
- apiVersion: v1
  kind: ConfigMap
  metadata:
    name: setters
    annotations:
      config.kubernetes.io/target-path: 'test/setters.yaml'
  data:
    namespace: config-control-2
```

**Pull request merged**
